lxd/api_cluster.go

```diff
@@ -2233,3 +2241,531 @@ func internalClusterRaftNodeDelete(d *Daemon, r *http.Request) response.Response
+func clusterNodeStatePost(d *Daemon, r *http.Request) response.Response {
+	// Parse the request
+	req := api.ClusterMemberStatePost{}
+	err := json.NewDecoder(r.Body).Decode(&req)
+	if err != nil {
+		return response.BadRequest(err)
+	}
+
+	if req.Action == "evacuate" {
+		return evacuateClusterMember(d, r)
+	} else if req.Action == "restore" {
+		return restoreClusterMember(d, r)
+	}
+
+	return response.BadRequest(fmt.Errorf("Unknown action %q", req.Action))
+}
+
+func evacuateClusterMember(d *Daemon, r *http.Request) response.Response {
+	sourceName := mux.Vars(r)["name"]
+
+	var node db.NodeInfo
+	var err error
+
+	err = d.cluster.Transaction(func(tx *db.ClusterTx) error {
+		// Get the node.
+		node, err = tx.GetNodeByName(sourceName)
+		if err != nil {
+			return errors.Wrap(err, "Failed to get node by name")
+		}
+
+		if node.State == db.ClusterMemberStatePending {
+			return fmt.Errorf("Cannot evacuate or restore a pending node")
+		}
+
+		// Set node status to EVACUATED to prevent instances from being created.
+		err = tx.UpdateNodeStatus(node.ID, db.ClusterMemberStateEvacuated)
+		if err != nil {
+			return errors.Wrap(err, "Failed to update node status")
+		}
+
+		return nil
+	})
+	if err != nil {
+		return response.SmartError(err)
+	}
+
+	// Do nothing if the node is already evacuated.
+	if node.State == db.ClusterMemberStateEvacuated {
+		return response.SmartError(fmt.Errorf("Node is already evacuated"))
+	}
+
+	reverter := revert.New()
+	defer reverter.Fail()
+
+	// Ensure node is put into its previous state if anything fails.
+	reverter.Add(func() {
+		d.cluster.Transaction(func(tx *db.ClusterTx) error {
+			tx.UpdateNodeStatus(node.ID, db.ClusterMemberStateCreated)
+
+			return nil
+		})
+	})
+
+	// The instances are retrieved in a separate transaction, after the node is in EVACUATED state.
+	var dbInstances []db.Instance
+
+	err = d.cluster.Transaction(func(tx *db.ClusterTx) error {
+		// If evacuating, consider only the instances on the node which needs to be evacuated.
+		dbInstances, err = tx.GetInstances(db.InstanceFilter{Node: sourceName})
+		if err != nil {
+			return errors.Wrap(err, "Failed to get instances")
+		}
+
+		return nil
+	})
+	if err != nil {
+		return response.SmartError(err)
+	}
+
+	instances := make([]instance.Instance, len(dbInstances))
+
+	for i, dbInst := range dbInstances {
+		inst, err := instance.LoadByProjectAndName(d.State(), dbInst.Project, dbInst.Name)
+		if err != nil {
+			return response.SmartError(errors.Wrap(err, "Failed to load instance"))
+		}
+
+		instances[i] = inst
+	}
+
+	var targetNodeName string
+	var targetNode db.NodeInfo
+
+	err = d.cluster.Transaction(func(tx *db.ClusterTx) error {
+		targetNodeName, err = tx.GetNodeWithLeastInstances([]int{node.Architecture}, -1)
+		if err != nil {
+			return err
+		}
+
+		targetNode, err = tx.GetNodeByName(targetNodeName)
+		if err != nil {
+			return err
+		}
+
+		return nil
+	})
+	if err != nil {
+		return response.SmartError(err)
+	}
```

Comment:
We need to do that query for every instance, otherwise all instances will be moved to the same target instead of getting properly spread on the remaining servers, potentially causing that one server to go out of resources and crash.